modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/ConnectionManager.java

```diff
@@ -398,117 +193,110 @@ public void stop() {
-            U.close(sysConn, log);
+        U.close(sysConn, log);
+    }
 
-            sysConn = null;
-        }
+    /**
+     * Called periodically to clean up the statement cache.
+     */
+    private void cleanupStatements() {
+        long now = U.currentTimeMillis();
+
+        connPool.forEach(c -> {
+            if (now - c.statementCache().lastUsage() > stmtTimeout)
+                c.clearStatementCache();
+        });
     }
 
     /**
-     * Handles SQL exception.
-     * @param c Connection to close.
+     * @param schema Schema name.
+     * @return Connection with setup schema.
      */
-    public void onSqlException(Connection c) {
-        H2ConnectionWrapper conn = threadConn.get().object();
+    public H2PooledConnection connection(String schema) {
+        H2PooledConnection conn = connection();
 
-        // Clear thread local cache if connection not detached.
-        if (conn.connection() == c)
-            threadConn.remove();
+        try {
+            conn.schema(schema);
 
-        if (c != null) {
-            threadConns.remove(Thread.currentThread());
+            return conn;
+        }
+        catch (IgniteSQLException e) {
+            U.closeQuiet(conn);
 
-            // Reset connection to receive new one at next call.
-            U.close(c, log);
+            throw e;
         }
     }
 
     /**
-     * Create new connection wrapper.
+     * Resize the connection pool.
      *
-     * @return Connection wrapper.
+     * @param size New size the connection pool.
      */
-    private H2ConnectionWrapper newConnectionWrapper() {
-        try {
-            return new H2ConnectionWrapper(DriverManager.getConnection(dbUrl));
-        }
-        catch (SQLException e) {
-            throw new IgniteSQLException("Failed to initialize DB connection: " + dbUrl, e);
-        }
+    void poolSize(int size) {
+        if (size <= 0)
+            throw new IllegalArgumentException("Invalid connection pool size: " + size);
+
+        connPool.resize(size);
     }
 
     /**
-     * Called by connection bool on connection recycle.
-     *
-     * @param conn recycled connection.
+     * @return H2 connection wrapper.
      */
-    private void addConnectionToThreaded(H2ConnectionWrapper conn) {
-        Thread cur = Thread.currentThread();
+    public H2PooledConnection connection() {
+        try {
+            H2Connection conn = connPool.borrow();
 
-        ConcurrentMap<H2ConnectionWrapper, Boolean> setConn = threadConns.get(cur);
+            if (conn == null)
+                conn = newConnection();
 
-        if (setConn == null) {
-            setConn = new ConcurrentHashMap<>();
+            H2PooledConnection connWrp = new H2PooledConnection(conn, this);
 
-            threadConns.putIfAbsent(cur, setConn);
-        }
+            usedConns.add(conn);
 
-        setConn.put(conn, false);
+            assert !conn.connection().isClosed() : "Connection is closed [conn=" + conn + ']';
```

Comment:
Is it possible to get connection from pool that has been closed due to some timeout?